Create test_wu2004.py

Jaydon2005 opened the pull request:
Creating unit tests to resolve an error found in `wu2004.ipynb`, documented in https://github.com/AgileBioFoundry/Bayesian-metabolic-control-analysis/issues/4
This test suite addresses a `ValueError` in the notebook caused by applying `pm.hdi()` directly to a pandas DataFrame. The fix wraps `pm.hdi()` in a helper (`series_to_hdi`) that converts Series to NumPy arrays before computing HDIs.

Tests included:
- `test_series_to_hdi_result_structure`: Ensures the wrapper returns a valid HDI Series.
- `test_apply_on_dataframe`: Confirms correct shape and type when applying across a DataFrame.

Run tests with:

```bash
    pytest test_wu2004.py
```

Dependencies: `pandas, numpy, pymc`

Jaydon2005 committed:
Create test_wu2004.py

Creating unit tests for wu2004.ipynb
This test suite addresses a ValueError in the notebook caused by applying `pm.hdi()` directly to a pandas DataFrame.
The fix wraps `pm.hdi()` in a helper (`series_to_hdi`) that converts Series to NumPy arrays before computing HDIs.

Tests included:
- `test_series_to_hdi_result_structure`: Ensures the wrapper returns a valid HDI Series.
- `test_apply_on_dataframe`: Confirms correct shape and type when applying across a DataFrame.

Run tests with:
    pytest test_wu2004.py

Dependencies: pandas, numpy, pymc, pytest

djinnome commented:
I looked it over. It looks great. Before I accept the pull request, a few more things:

- [x] Remove the function `series_to_hdi` from `test_wu2004.py` and add it to `src/emll/util.py`
- [x]  Remove lines [43-46](https://github.com/AgileBioFoundry/Bayesian-metabolic-control-analysis/pull/5#discussion_r2162852443)
- [x] Remove lines [22-26](https://github.com/AgileBioFoundry/Bayesian-metabolic-control-analysis/pull/5#discussion_r2162851758)
- [x] Remove lines [6-10](https://github.com/AgileBioFoundry/Bayesian-metabolic-control-analysis/pull/5#discussion_r2162851758) 
- [x] Rename the file `test_wu2004.py` and call it `test_util.py`
- [x] Inside `test_util.py`  add from `emll.util import series_to_hdi`
- [x] In the beginning of `wu2004.ipynb`  add `from emll.util import series_to_hdi`
- [x] Replace the cell that calls `e_df.apply( pm.hdi )` with `e_df.appy(series_to_hdi)`
- [x] Make sure `pytest test_util.py` passes.
- [x] Make sure that you can execute the full notebook `wu2004.ipynb` without any errors.

djinnome commented on tests/test_wu2004.py:
```diff
@@ -0,0 +1,46 @@
+    hdi_bounds = pm.hdi(arr, hdi_prob=hdi_prob)
+    return pd.Series(hdi_bounds, index=["hdi_lower", "hdi_upper"])
+
+# --- Usage: apply across dataframe columns ---
```
Remove lines 22-26

djinnome replied:
Remove lines 22-26

djinnome commented on tests/test_wu2004.py:
```diff
@@ -0,0 +1,46 @@
+    assert isinstance(out, pd.DataFrame), "Output should be a DataFrame"
+    assert out.shape == (2, 3), "Result DataFrame must have 2 rows (bounds) x n_columns"
+
+# Optionally: run tests immediately in notebook
```
Remove lines 43-46

djinnome replied:
Remove lines 43-46

djinnome commented on tests/test_wu2004.py:
```diff
@@ -0,0 +1,46 @@
+import pymc as pm
+import pytest
+
+# --- Sample / Real data setup (replace e_all and e_labels with your real data) ---
```
Remove lines 6-10

djinnome replied:
Remove lines 6-10

djinnome commented on tests/test_wu2004.py:
```diff
@@ -0,0 +1,46 @@
+e_df = pd.DataFrame(e_all, columns=e_labels)
+
+# --- Fix: wrapper converting pandas Series to numpy array before HDI ---
+def series_to_hdi(series: pd.Series, hdi_prob: float = 0.94) -> pd.Series:
```
Move  `series_to_hdi` to `src/emll/util.py` and on line 5 in this file, add:
```python
from emll.util import series_to_hdi
```

djinnome reviewed (changes requested):
- [x] Change the name of this file to `test_util.py`.  
- [x] update `wu2004.ipynb` by using `e_df.apply(series_to_hdi)` instead of `e_df.apply(pm.hdi)`

Jaydon2005 committed:
Removed the function series_to_hdi from test_wu2004.py and add it to src/emll/util.py

Added function series_to_hdi at the end of src/emll/util.py. Also added import pandas as pd

Jaydon2005 committed:
Update and rename test_wu2004.py to test_util.py

Removed lines 6-10, 22-26, 43-46
Renamed the file test_wu2004.py and updated to test_util.py
Inside test_util.py add from emll.util import series_to_hdi
In the beginning of wu2004.ipynb add from emll.util import series_to_hdi
Removed series_to_hdi function (lines 13–20).

Jaydon2005 committed:
Update wu2004.ipynb

Replaced e_hdi = e_df.apply(pm.hdi) to e_hdi = e_df.apply(series_to_hdi)

Jaydon2005 committed:
Update wu2004.ipynb

Forgot to add "e_hdi =" next to "e_df.apply(series_to_hdi)"

Jaydon2005 committed:
Added from emll.util import series_to_hdi

After ln [3] where seaborn is imported from emll.util import series_to_hdi is imported

Jaydon2005 committed:
Update wu2004.ipynb Added more imports to fix ImportError: cannot import name 'series_to_hdi' from 'emll.util' (/Users/lari627/bayesian-metabolic-control-analysis/src/emll/util.py)

Added the following imports:
import matplotlib.pyplot as plt
import importlib
import emll.util
importlib.reload(emll.util)

So now it looks like:
from tqdm import tqdm

import matplotlib.pyplot as plt
import importlib
import emll.util
importlib.reload(emll.util)

from emll.util import series_to_hdi
import seaborn as sns
sns.set(context='talk', style='ticks',
        color_codes=True, rc={'legend.frameon': False})

Jaydon2005 committed:
Fixed spaces "\n", on line 67 and 71

FIxed sp

Jaydon2005 committed:
Update wu2004.ipynb Removed "importlib.reload(emll.util)\n"

Jaydon2005 committed:
Added more imports 

Added the following imports below:
import importlib
import emll.util
importlib.reload(emll.util)

Jaydon2005 committed:
Removed imports importlib and emll.util

Moved from emll.util import series_to_hdi to e_df = pd.DataFrame(e_all, columns=e_labels)
e_df.apply(series_to_hdi)
